File: auth/src/main/java/com/sherpa/auth/user/usecase/OAuthSignInUseCase.java
package com.sherpa.auth.user.usecase;

import org.springframework.stereotype.Service;

import com.sherpa.auth.user.dto.OAuthLoginCommand;
import com.sherpa.auth.user.dto.OAuthRegisterMemberCommand;
import com.sherpa.auth.user.dto.OAuthSignInCommand;
import com.sherpa.auth.user.dto.OAuthSignInResult;
import com.sherpa.auth.user.service.OAuthLoginService;
import com.sherpa.auth.user.service.OAuthRegisterMemberService;

@Service
public class OAuthSignInUseCase {

	private final OAuthRegisterMemberService oAuthRegisterMemberService;
	private final OAuthLoginService oAuthLoginService;

	public OAuthSignInUseCase(OAuthRegisterMemberService oAuthRegisterMemberService,
		OAuthLoginService oAuthLoginService) {
		this.oAuthRegisterMemberService = oAuthRegisterMemberService;
		this.oAuthLoginService = oAuthLoginService;
	}

	public OAuthSignInResult oAuthSignIn(OAuthSignInCommand signInCommand) {

		var registerCommand = OAuthRegisterMemberCommand.builder()
			.name(signInCommand.getName())
			.email(signInCommand.getEmail())
			.profileUrl(signInCommand.getProfileUrl())
			.build();
		oAuthRegisterMemberService.registerOAuthMember(registerCommand);

		var oAuthLoginCommand = OAuthLoginCommand.builder()
			.email(signInCommand.getEmail())
			.build();

		var oAuthLoginResult = oAuthLoginService.login(oAuthLoginCommand);

		return OAuthSignInResult.builder()
			.accessToken(oAuthLoginResult.getAccessToken())
			.build();
	}

}

File: gateway/src/main/kotlin/com/sherpa/gateway/domain/auth/usecase/OAuthSignInUseCase.kt
package com.sherpa.gateway.domain.auth.usecase

import com.sherpa.gateway.domain.auth.client.AuthHttpClient
import com.sherpa.gateway.domain.auth.dto.OAuthRegisterMemberRequest
import com.sherpa.gateway.domain.auth.dto.OAuthSignInCommand
import com.sherpa.gateway.domain.auth.dto.OAuthSignInResult
import com.sherpa.gateway.domain.auth.jwt.AccessTokenPayload
import com.sherpa.gateway.domain.auth.jwt.JwtService
import com.sherpa.gateway.domain.auth.oauthattributes.OAuthAttributes
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import java.util.*

@Service
class OAuthSignInUseCase(
    private val authHttpClient: AuthHttpClient,
    private val jwtService: JwtService
) {

    suspend fun sendOAuthRegisterRequest(attributes: OAuthAttributes): ResponseEntity<Void> {
        val response = authHttpClient.postOAuthMember(
            OAuthRegisterMemberRequest(
                attributes.email,
                attributes.name,
                attributes.profileURL
            )
        )
        return response
    }

    fun createAccessToken(email: String): String {
        val payload = AccessTokenPayload(
            email,
            issuedAt = Date(System.currentTimeMillis())
        )
        return jwtService.createAccessToken(payload)
    }

    suspend fun oAuthSignIn(command: OAuthSignInCommand): OAuthSignInResult {
        val attributes = command.attributes
        sendOAuthRegisterRequest(attributes)
        val accessToken = createAccessToken(attributes.email)

        return OAuthSignInResult(accessToken)
    }


}